### PR TITLE
PLT-6403: Interactive messages

### PR DESCRIPTION
**Pull request by ccbrown**
#### Summary
Basic interactive message buttons. Inspired primarily by [Outlook's card actions](https://docs.microsoft.com/en-us/outlook/actionable-messages/card-reference#actions).

[Demo](https://gfycat.com/ShamefulHandsomeBlackwidowspider)

This allows you to add "actions" to post attachments:

```json
{
  "attachments": [
    {
      "text": "This is the attachment text.",
      "actions": [
        {
          "name": "Ephemeral Message",
          "integration": {
            "url": "http://127.0.0.1:7357",
            "context": {
              "action": "ephemeral"
            }
          }
        }, {
          "name": "Update",
          "integration": {
            "url": "http://127.0.0.1:7357",
            "context": {
              "action": "update"
            }
          }
        }
      ]
    }
  ]
}
```

The actions are backed by an integration that will handle POST requests when users click the action buttons. These requests contain the user id and any context that was provided in the action definition:

```json
{
  "user_id": "rd49ehbqyjytddasoownkuqrxe",
  "context": {
    "action": "ephemeral"
  }
}
```

The integration can respond with an update to the original post and/or an ephemeral message:

```json
{
  "update": {
    "message": "Updated!"
  },
  "ephemeral_text": "You updated the post!"
}
```

Things for future iterations:

* Button styles such as "dangerous" buttons.
* Client-side action types such as "open URL in browser".
* Actions with additional inputs such as text fields.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6403

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

### CODE REVIEW

**Comment by jasonblais** [comment]
Sweeeet.. Thanks @ccbrown. Few questions:

1) Does the above JSON payload contain all the features we currently support for the interactive messages?
2) `The integration can respond with an update to the original post` --> what responses are supported? In addition to the ephemeral message action.
3) On naming of the payload:

```
        {
          "name": "Ephemeral Message",
          "integration": {
            "url": "http://127.0.0.1:7357",
            "context": {
              "action": "ephemeral"
            }
          }
        }
```

3a) Why `integration`? Wondering if that's best choice? What other alternatives have we considered?
3b) Why `url` instead of `target`? 
3c) Are some of the additional fields Slack support relevant for this case? They have fields such as `callback_id` and `fallback`
4) Can we get an example of polling on the spinmint?
5) I'll also ask Asaad's help with CSS/styling review later this week.

**Comment by ccbrown** [comment]
@jasonblais 

1. Currently, yes.
2. It can replace the original message, add an ephemeral message, or both. Which is slightly more than what both Slack integrations and Outlook actions can do.
3. So first of all, I didn't want to do this quite like Outlook does with the `@type` field for two reasons:
    * I think creating a hierarchical schema is generally better than having to write things like "this field is ignored if the value of the type field is not HttpPost" in the documentation for all of your other fields.
    * I wanted to be able to support fallbacks by allowing multiple alternative action outcomes. For example, say we add an action type that opens a URL in the browser. Then we release a version of the app with that feature. Then down the road we add an action type that prompts for additional input. Integration developers might want to define actions that require additional input such that if the user's app supports it, users provide the input in-app, but otherwise, the button just opens a web page in the browser where the user can perform the same action outside of the app.

    I chose "integration" because it seemed like a fitting noun to use to define the action handler and is one that we use commonly for this sort of thing. I also wanted to capture the fact that because the response has meaning specific to Mattermost, this is something slightly different than just a generic HTTP POST (e.g. you probably shouldn't configure your action to POST straight to the GitHub API). Don't have any alternatives off the top of my head.

    The Outlook documentation describes the `target` field as "Defines the URL endpoint of the service that implements the action.". Considering "target" is kinda vague and they themselves describe it as "the URL", it seemed more logical to me to just name the field "url".

    The Slack `callback_id` is necessary for them because your app only defines a single URL for all actions. The integration that gets invoked when the action is triggered wouldn't have any way of knowing _which_ action was triggered without this. With the integration being configured per-post instead of per-app, you can easily define an equivalent in the URL itself or in the `context` field.

    The Slack `fallback` field isn't specific to actions (and we already support it). But for action fallbacks, see above.

4. Yeah I'll set something up today.

5. It's probably better to do them like Slack, where instead of specifying a color for the button, you specify something like a `"style": "danger"` field. That way they'll stand out appropriately regardless of which theme is being used.

**Comment by jasonblais** [comment]
Thanks @ccbrown! All sounds good to me. Great work on this.

**Comment by ccbrown** [comment]
@jasonblais Simple polling integration set up on the test instance:

[Demo](https://gfycat.com/UntimelyPoisedAnkole)

The code for that can be found here: https://github.com/ccbrown/interactive-post-demo

I think this is a solid building block, but there are several additional features that would make this really good. For example, for the polling demo, it would have been nice if I could have made that "End Poll" button visible only to users with edit permissions for the post. And of course, the things listed in PR description and our design documents.

**Comment by jasonblais** [comment]
Thanks @ccbrown! The CSS and styling got reviewed, so we should be fine on that front. Have also being testing different UI and looks good. Included desktop on Win 10 + mobile browser and the React Native app on iOS. It doesn't work on the React Native apps as one would expect, but we can queue the support for it later.

Some more questions, thank you in advance:

1) For polling, is there actually a way to do proper polling with the buttons? Because the sample polling integration is okay, but it won't really be what customers are looking for (If it's something that the interactive buttons would support, but the quick integration you built doesn't, that's okay).

If it doesn't, the Outlook cards have the `MultichoiceInput` feature as part of the cards, where you can supply the choices through the webhook and display as a poll.

image
image

PS: I definitely don't want to introduce a feature creep here. Several customers have asked for polling support though, so just thinking if there's an opportunity for it here.

2) When there is an action on the button, the messages posted to the channel aren't displayed as replies, because they're ephemeral messages, but do increase the reply count on the original post. 

Is that intended?

image

3) For the "update" action, I assume the `(edited)` indicator is shown as by nature the user is editing the original post?

image

4) Once a user hits the "Update" button, does it update the post to everyone or only themselves?

**Comment by ccbrown** [comment]
@jasonblais 

1. You could make the buttons do or say whatever you like. So you could do more advanced polling with several more, customized options, but you're limited to what you can do with buttons at the moment. I think as far as polling goes, this feature is already just as capable as it would be if it had a "multi-select" input. The UX would just be better with additional features like that.

2. This is intentional. It's so the ephemeral messages appear in the RHS as well, and it's the same behavior as when you perform a slash command in the RHS that creates an ephemeral message.

3. Yep.

4. Everyone.

**Comment by jasonblais** [comment]
Thanks @ccbrown 

1. Okay cool. Last question: Is there enough support here where someone can build an integration that mimics Slack polls like https://pluspackagetesting.slack.com/apps/A0HFW7MR6-simple-poll

2. Oh I didn't realize we updated the reply count when you perform a slash command in the RHS. It kind of makes sense I suppose, but not 100% sure..either way, it's consistent here.

3. From what I understand, there would be a way to build an integration to show who clicked the "Update" button, like attached. Is that correct?

image

4. Thanks. Sounds good.

I'll direct further questions to the Dev Toolkit channel so not to delay this PR. Just questions that come to mind as I'm thinking of customer questions and the upcoming documentation,

**Comment by ccbrown** [comment]
@jasonblais 

1. Yes, you could do that.
3.  Yeah, you could do that too.

There are two things I might want to consider adding to make these cases better though:

* I believe there's a race condition present in most Slack apps (and probably Outlook cards) that do the sort of polling in 1. If two users click a button at the same time, both of their votes may not be reflected in the updated post. A concept of post revision numbers to eliminate that race condition might be nice.
* Maybe the clicker's user name should be included in the POST body? Currently, the integration would have to look up the user name via the API.

**Comment by jasonblais** [comment]
> There are two things I might want to consider adding to make these cases better though:

I presume these improvements/changes would be in a separate PR later?

**Comment by ccbrown** [comment]
@jasonblais Yeah, considering Slack and Outlook don't have those things, it seems like they can wait until the next batch of improvements.

File: api4/post.go [hunk 1]
```diff
@@ -428,3 +430,25 @@ func getFileInfosForPost(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if !app.SessionHasPermissionToChannelByPost(c.Session, c.Params.PostId, model.PERMISSION_READ_CHANNEL) {
+		c.SetPermissionError(model.PERMISSION_READ_CHANNEL)
+		return
+	}
+
+	params := mux.Vars(r)
+	id := params["id"]
```
Comment (jwilander) [inline api4/post.go]: Can you add this to `api4/params.go` and add a context function `RequireActionId()`? Then you can just do `c.RequirePostId().RequireActionId()` at the top here and for any future functions that need an action id

File: webapp/actions/post_actions.jsx [hunk 2]
```diff
@@ -351,3 +351,9 @@ export function unpinPost(postId) {
         });
     };
 }
+
+export function doPostAction(postId, actionId) {
+    Client4.doPostAction(postId, actionId).catch(
+        () => {} //eslint-disable-line no-empty-function
+    );
```
Comment (jwilander) [inline webapp/actions/post_actions.jsx]: This really should be a mattermost-redux action so it can be used by RN

File: webapp/package.json [hunk 3]
```diff
@@ -25,7 +25,7 @@
     "localforage": "1.5.0",
     "marked": "mattermost/marked#5194fc037b35036910c6542b04bb471fe56b27a9",
     "match-at": "0.1.0",
-    "mattermost-redux": "mattermost/mattermost-redux#webapp-4.1",
+    "mattermost-redux": "ccbrown/mattermost-redux#PLT-6403",
```
Comment (jwilander) [inline webapp/package.json]: This will need to be updated to latest master once it's merged into mattermost-redux

**Comment by ccbrown** [comment]
@jwilander All updated (except the package.json change, which I'll update after the redux PR is merged).

**Comment by ccbrown** [comment]
Build failure is just enterprise being out of sync.

File: app/post.go [hunk 4]
```diff
@@ -632,3 +635,65 @@ func GetOpenGraphMetadata(url string) *opengraph.OpenGraph {
...
+	}
+
+	action := post.GetAction(actionId)
+	if action == nil || action.Integration == nil {
+		return model.NewAppError("DoPostAction", "api.post.do_action.action_id.app_error", nil, fmt.Sprintf("action=%v", action), http.StatusNotFound)
+	}
+
+	request := &model.PostActionIntegrationRequest{
+		UserId:  userId,
+		Context: action.Integration.Context,
```
[the remark this answers is not in the document]
Comment (ccbrown) [reply app/post.go]: It's defined by the thing that made the post and actually can be anything. It's used by the post creator to store arbitrary data such as a secret for authenticating Mattermost or a poll id or whatever. Outlook has the same sort of thing, but is restricted to string -> string mappings and passed as headers. It's very handy.

File: app/post.go [hunk 5]
```diff
@@ -632,3 +635,65 @@ func GetOpenGraphMetadata(url string) *opengraph.OpenGraph {
...
+		response.Update.Id = postId
+		response.Update.AddProp("from_webhook", "true")
+		if _, err := UpdatePost(response.Update, false); err != nil {
+			return err
+		}
+	}
+
+	if response.EphemeralText != "" {
+		ephemeralPost := &model.Post{}
+		ephemeralPost.Message = parseSlackLinksToMarkdown(response.EphemeralText)
```
[the remark this answers is not in the document]
Comment (ccbrown) [reply app/post.go]: /shrug Also 0/5.

File: app/post.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/post.go]: Okay cool, that makes sense then

**Comment by ccbrown** [comment]
@quadriq Not at the moment, but I do think this would be a nice feature to have.

The only reason I haven't added it already is because it would introduce security concerns; we would need users to be able to see what slash command they're going to be triggering beforehand. Otherwise, I'd never be able to click an action button out of fear of it executing slash commands I don't want it to.

Feel free to request it and maybe we'll prioritize it higher: https://www.mattermost.org/feature-ideas/

**Pull request merged**